src/nccl_ofi_rdma.c

```diff
@@ -1438,22 +1313,125 @@ static inline int process_completions(struct fi_cq_tagged_entry *cq_entry,
-			ret = handle_bounce_recv(&cq_entry[comp_idx], rail->rail_id);
+			/* Receive completions */
+			assert(NULL != req);
+
+			assert(req->type == NCCL_OFI_RDMA_BOUNCE);
+
+			if (!(comp_flags & FI_REMOTE_CQ_DATA)) {
+				/* CONN, CONN_RESP, or CTRL message */
+				msg_type = (nccl_ofi_rdma_msg_type_t *)cq_entry[comp_idx].buf;
+
+				if (*msg_type == NCCL_OFI_RDMA_MSG_CONN) {
+					/* CONN receive completion */
+					assert(cq_entry[comp_idx].len == sizeof(nccl_ofi_rdma_connection_info_t));
+
+					conn_msg = (nccl_ofi_rdma_connection_info_t *)cq_entry[comp_idx].buf;
+					l_comm = (nccl_net_ofi_rdma_listen_comm_t *)get_comm(ep,
+											     conn_msg->remote_comm_id);
+
+					assert(l_comm->base.base.type == NCCL_NET_OFI_LISTEN_COMM);
+					assert(l_comm->req.comm->type == NCCL_NET_OFI_LISTEN_COMM);
+					assert((nccl_net_ofi_comm_t *)l_comm == l_comm->req.comm);
+
+					/* Copy connection message in the communicator */
+					l_comm->conn_msg = *conn_msg;
+
+					ret = inc_req_completion(&l_comm->req, cq_entry[comp_idx].len, 1);
+					if (OFI_UNLIKELY(ret != 0)) {
+						goto exit;
+					}
+
+					/* Attempt to re-post bounce buffer */
+					ret = repost_bounce_buff(ep, req);
+					if (OFI_UNLIKELY(ret != 0)) {
+						NCCL_OFI_WARN("Failed to repost bounce buff");
+						goto exit;
+					}
+				} else if (*msg_type == NCCL_OFI_RDMA_MSG_CONN_RESP) {
+					/* CONN_RESP receive completion */
+					assert(sizeof(nccl_ofi_rdma_connection_info_t) == cq_entry[comp_idx].len);
+
+					conn_resp_msg = (nccl_ofi_rdma_connection_info_t *)cq_entry[comp_idx].buf;
+					s_comm = (nccl_net_ofi_rdma_send_comm_t *)get_comm(
+						ep, conn_resp_msg->remote_comm_id);
+
+					assert(NCCL_NET_OFI_SEND_COMM == s_comm->base.base.type);
+					assert(NULL != s_comm->conn_resp_req);
+					assert(NCCL_NET_OFI_SEND_COMM == s_comm->conn_resp_req->comm->type);
+					assert((nccl_net_ofi_comm_t *)s_comm == s_comm->conn_resp_req->comm);
+
+					/* Copy connection response message in the communicator */
+					s_comm->conn_msg = *conn_resp_msg;
+
+					ret = inc_req_completion(s_comm->conn_resp_req, cq_entry[comp_idx].len, 1);
+					if (OFI_UNLIKELY(ret != 0)) {
+						goto exit;
+					}
+
+					ret = finish_connect(s_comm);
+					if (OFI_UNLIKELY(ret != 0)) {
+						goto exit;
+					}
+
+					/* Attempt to re-post bounce buffer */
+					ret = repost_bounce_buff(ep, req);
+					if (OFI_UNLIKELY(ret != 0)) {
+						NCCL_OFI_WARN("Failed to repost bounce buff");
+						goto exit;
+					}
+				} else if (*msg_type == NCCL_OFI_RDMA_MSG_CTRL) {
+					/* CTRL receive completion */
+					assert(sizeof(nccl_net_ofi_rdma_ctrl_msg_t) == cq_entry[comp_idx].len);
+
+					ctrl_msg = (nccl_net_ofi_rdma_ctrl_msg_t *)cq_entry[comp_idx].buf;
+
+					s_comm =
+						(nccl_net_ofi_rdma_send_comm_t *)get_comm(ep, ctrl_msg->remote_comm_id);
+
+					assert(NCCL_NET_OFI_SEND_COMM == s_comm->base.base.type);
+
+					get_bounce_data(req)->recv_len = cq_entry[comp_idx].len;
```

Comment:
This line is duplicated below, which is error-prone. Consider having a `handle_bounce_req`, like before, which sets this `recv_len` member and then has cases for the different bounce recv types.